### PR TITLE
Add VolumeDataPopulated condition

### PR DESCRIPTION
**Pull request by jingxu97**
address design proposal: https://github.com/kubernetes/community/pull/2495
    
This change add a condition called "VolumeDataPopulated" into PV. With the feature of snapshot and data source, volume might be created from snapshot source with data populated into the volume instead of an empty one. The external provision needs to understand the data source and also mark the condition "VolumeDataPopualted" to true. The PV controller will check this condition if the data source of PVC is specified. Only if the condition is true, the PVC can be marked as "Bound" phase.

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -350,6 +350,29 @@ type PersistentVolumeStatus struct {
 	// Reason is a brief CamelCase string that describes any failure and is meant for machine parsing and tidy display in the CLI
 	// +optional
 	Reason string
+	// +optional
+	Conditions []PersistentVolumeCondition
+}
+
+type PersistentVolumeConditionType string
+
+// These are valid conditions of Pvc
```
Comment (xing-yang) [inline pkg/apis/core/types.go]: s/Pvc/PV

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -380,6 +380,29 @@ type PersistentVolumeStatus struct {
 	// for machine parsing and tidy display in the CLI.
 	// +optional
 	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+	// +optional
+	Conditions []PersistentVolumeCondition
+}
+
+type PersistentVolumeConditionType string
+
+// These are valid conditions of Pvc
```
Comment (xing-yang) [inline staging/src/k8s.io/api/core/v1/types.go]: s/Pvc/PV

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 3]
```diff
@@ -1012,6 +1012,11 @@ func (ctrl *PersistentVolumeController) bind(volume *v1.PersistentVolume, claim
 	}
 	claim = updatedClaim
 
+	if claim.Spec.DataSource != nil && !isVolumeDataPopulated(volume) {
+		errStr := fmt.Sprintf("error binding volume %q to claim %q: claim has data source, but data is not yet populated to volume", volume.Name, claimToClaimKey(claim))
+		glog.V(2).Infof(errStr)
```
Comment (neolit123) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: should be `V(3)`?


---
something else to note - i think the error bubbles up to the PVC bind() callers, yet Infof() is called both here and in the callers.
ideally the pattern for using glog should be calling Infof() once, but i see the same for other logic in this function...so this change is conformant.

**Comment by neolit123** [comment]
please run these locally and make sure they pass:
```
./hack/verify-bazel.sh
./hack/verify-gofmt.sh
./hack/verify-golint.sh
```

**Comment by neolit123** [comment]
this probably needs a release note.
/sig apps
/kind api

**Comment by xing-yang** [comment]
Talked to @jingxu97.  She asked me to add code from this PR to https://github.com/kubernetes/kubernetes/pull/67625.

**Pull request closed without merging**